### ISSUE TITLE
Unable to connect to container or do port-forward while kubectl is behind a proxy

### ISSUE THREAD

**Issue by luisyonaldo** [kubernetes/kubernetes#58065]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
 /kind bug
> /kind feature

**What happened**:
I can't connect to a pod (`kubectl exec -it`) or do a port forwarding (`kubectl port-porward`) behind a proxy.

**What you expected to happen**:
Connect to the pod container and  be able to do a port-forward (this is mostly for be able to run helm locally and communicate with tiller running inside the cluster).

**How to reproduce it (as minimally and precisely as possible)**:
I'm using a socks proxy to communicate with cluster nodes. I set all environment variables to use proxy (`http_proxy`, `https_proxy` and `no_proxy`) for both `docker` and `kubectl`. I can communicate with the cluster using kubectl (so `kubectl get node`, `kubectl create`, etc) but I can't connect to a container or do port-forward.

```
$ kubectl exec -it prometheus-kube-prometheus-0 bash -n monitoring
Defaulting container name to prometheus.
Use 'kubectl describe pod/prometheus-kube-prometheus-0 -n monitoring' to see all of the containers in this pod.
error: error sending request: Post https://<server ip>/api/v1/namespaces/monitoring/pods/prometheus-kube-prometheus-0/exec?command=bash&container=prometheus&container=prometheus&stdin=true&stdout=true&tty=true: EOF
```
same with
```
$ kubectl port-forward tiller-deploy-546cf9696c-qx5vx 4134 -n kube-system
error: error upgrading connection: error sending request: Post https://<server ip>/api/v1/namespaces/kube-system/pods/tiller-deploy-546cf9696c-qx5vx/portforward: EOF
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1", GitCommit:"3a1c9449a956b6026f075fa3134ff92f7d55f812", GitTreeState:"clean", BuildDate:"2018-01-04T20:00:41Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.1+coreos.0", GitCommit:"59359d9fdce74738ac9a672d2f31e9a346c5cece", GitTreeState:"clean", BuildDate:"2017-10-12T21:53:13Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
- OS (e.g. from /etc/os-release):
```
NAME="Red Hat Enterprise Linux Server"
VERSION="7.3 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="7.3"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.3 (Maipo)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.3:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.3
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.3"
```
- Kernel (e.g. `uname -a`):
```
Linux tapps773 3.10.0-514.26.1.el7.x86_64 #1 SMP Tue Jun 20 01:16:02 EDT 2017 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools:
kubespray / kubeadmin

**Comment by ncdc** [comment]
exec and port-forward do support using a proxy using the http/https `CONNECT` mechanism. I'm thinking that maybe SOCKS is the issue? cc @liggitt

**Comment by luisyonaldo** [comment]
I also tunneled socks proxy to a http proxy but now kubectl doesn't work.

```
$ kubectl get nodes
Unable to connect to the server: Forbidden port
$ curl https://<api-server> -k
curl: (56) Received HTTP code 403 from proxy after CONNECT

```
as you can see is using `CONNECT` mechanism but still doesn't work.

**Comment by stepanstipl** [comment]
We hit the same issue.

Seems like "normal" http rest based kubectl commands (like get) work fine with Socks proxy, but the ones based on SPDY2 fail (attach, exec, port-forward). There's general move towards Websockets, I believe this is already implemented on server side (#13885) and there's work being done to move kubectl to Websockets too (#48641, #50428). I believe that once this lands, the resp. commands will work with Socks proxy too.

Cced @smarterclayton (as he seems to be the one behind most of the spdy2 -> websockets work)

**Comment by andelhie** [comment]
just ran in to this problem now to my self.  Is there a work around for this yet?

**Comment by KJTsanaktsidis** [comment]
This is affecting me as well. I did a packet capture, and what I found with `kubectl exec` was that:

* It made an initial GET request for the Pod over the SOCKS proxy just fine
* Then opens up a second TCP connection to the SOCKS proxy (so far so good)
* Then tries to do a TLS handshake on this socket, without first sending a SOCKS connect request through it

At a guess, it's forgotten that the proxy is socks5 and not https for the second connection?

**Comment by qzio** [comment]
Just want to report that I get similar problem using tinyproxy.

`curl: (56) Received HTTP code 403 from proxy after CONNECT`

However, using squid with the default configuration and
`HTTP_PROXY=<proxy-ip-and-port> HTTPS_PROXY=<proxy-ip-and-port> kubectl get pods` I get through without any error

**Comment by kfox1111** [comment]
I use socks5 proxies all the time via ssh, and it would be great if I could use it with kubectl.

**Comment by hshang09** [comment]
Got the same issue here. Any work around here?

**Comment by 11xor6** [comment]
So I'm encountering this issue using an SSH proxy (using `-D [port_number]`) and I've worked around this by installing an HTTP proxy (tinyproxy in my case) on the remote machine and then using port forwarding instead (`-L-L8888:127.0.0.1:8888`). 

A similar workaround in case you don't control the SOCKS proxy or don't use SSH for your proxy is to use a local HTTP(S) proxy which supports a SOCKS proxy as it's target as noted [here](https://superuser.com/questions/280129/http-proxy-over-ssh-not-socks). I haven't tested this method, but it seems like a solid solution.

**Comment by sjmiller609** [comment]
> So I'm encountering this issue using an SSH proxy (using `-D [port_number]`) and I've worked around this by installing an HTTP proxy (tinyproxy in my case) on the remote machine and then using port forwarding instead (`-L-L8888:127.0.0.1:8888`).
> 
> A similar workaround in case you don't control the SOCKS proxy or don't use SSH for your proxy is to use a local HTTP(S) proxy which supports a SOCKS proxy as its target as noted [here](https://superuser.com/questions/280129/http-proxy-over-ssh-not-socks). I haven't tested this method, but it seems like a solid solution.

For my use case, I am attempting to use a bastion host authenticated with google cloud IAP to connect to a private kubernetes endpoint. @11xor6 's comment works for me.

In on the bastion, you only need to run (example - ubuntu)
```
sudo apt-get -y update && sudo apt-get install -y tinyproxy
```
You should be able to start the port forward to access the proxy like this (access on local port 1234)
```
gcloud beta compute ssh --zone ${google_compute_instance.bastion.zone} ${google_compute_instance.bastion.name} --tunnel-through-iap --ssh-flag='-L 1234:127.0.0.1:8888 -C -N'
```
then this command should work
```
HTTPS_PROXY=http://127.0.0.1:1234 helm list
```

**Comment by binlab** [comment]
Any ideas for fix working kubectl with a socks5 proxy?

**Comment by terinjokes** [comment]
For non-SPDY requests, `kubectl` uses the proxying support in Go's net/http without modification. This works correctly for SOCKS5 proxies.

For SPDY connections, there's a separate RoundTripper in the [k8s.io/apimachinery/pkg/util/httpstream/spdy](https://github.com/kubernetes/kubernetes/blob/044a4ce21de7e89308be6cedc59fbf1cbb6b30b9/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go#L118) package. If I'm understanding the code correctly, it's assuming that if a proxy is configured, it's an HTTP or HTTPS proxy, and attempts to setup the proxy with the CONNECT method.

---

The difference is very evident in a tcpdump of the connection between `kubectl` and the SOCKS5 proxy. For an `kubectl exec` we can see two requests in the dump.

```console
$ env HTTPS_PROXY=socks5://127.0.0.1:1337 kubectl -n default exec shell-47h9q date
I0324 20:35:25.751847 1307825 loader.go:375] Config loaded from file:  /home/terin/.kube/config
I0324 20:35:25.984289 1307825 round_trippers.go:443] GET https://kubernetes.local/api/v1/namespaces/default/pods/shell-47h9q 200 OK in 218 milliseconds
I0324 20:35:26.006706 1307825 round_trippers.go:443] POST https://kubernetes.local/api/v1/namespaces/default/pods/shell-47h9q/exec?command=date&container=shell&stderr=true&stdout=true  in 1 milliseconds
F0324 20:35:26.006800 1307825 helpers.go:115] error: error sending request: Post https://kubernetes.local/api/v1/namespaces/default/pods/shell-47h9q/exec?command=date&container=shell&stderr=true&stdout=true: read tcp 127.0.0.1:39334->127.0.0.1:1337: read: connection reset by peer

```

![2020-03-24-20:27:03-screenshot](https://user-images.githubusercontent.com/273509/77498856-e2fbbb80-6e0d-11ea-93f5-065f4538f8de.png)

The first connection is `kubectl` making a GET request for information about the pod. In packets 4 and 6 we can see proxy authentication being negotiated and accepted, followed by packets 8 and 10 setting up the upstream connection and the success message. Standard application traffic (in this case, the TLS handshake followed by HTTP request and response) follows on the same connection.

The second connection is `kubectl` attempting to setup the SPDY connection for the exec command. The first message from the client is not setting up the SOCKS5 proxy, but instead jumps ahead entirely to the TLS handshake. The SOCKS5 proxy, rightly, resets the connection.

---

I note that [Go's implementation](https://github.com/golang/go/blob/20a838ab94178c55bc4dc23ddc332fce8545a493/src/net/http/transport.go#L1539-L1556) switches on the proxy URL scheme to chose different behaviors. It looks like would need to add an implementation for SOCKS5 (it's unfortunate the one in net/http is unexported).

**Comment by mulatinho** [comment]
I would love to get this feature, same problem here.

**Comment by mulatinho** [comment]
#72943 this was created but not merged :/ SOCKS5 is really useful and HTTP proxy keeps closing connection after a simple `kubectl logs`

**Comment by TGPSKI** [comment]
@mulatinho: here's the active PR: https://github.com/kubernetes/kubernetes/pull/84205

There's still a few open questions since last fall and I haven't gotten a response from the reviewers.

**Comment by tpoindessous** [comment]
/remove-lifecycle stale

thanks

**Comment by tpoindessous** [comment]
Work In Progress in https://github.com/kubernetes/kubernetes/pull/84205

**Comment by leonK-BO** [comment]
Hello,

I have a similiar issue.
My client is behind a corporate proxy and most of the kubectl commands work fine. But if I try to exec into a pod shell than I get following error:
```
kubectl exec -i -t -n default web-6798b588cf-4gzvq -c web "--" sh -c "clear; (bash || ash || sh)"
error: unable to upgrade connection: error dialing backend: dial tcp: lookup .....eks.amazonaws.com on 10.58.194.16:53: no such host
```

So kubectl tries to resolute the public EKS domain with our internal DNS. Our internal DNS does not support resolution of public domains. It looks like the similiar root cause for me. Do you know if there is a workaround for that or will this bug be fixed soon?

**Comment by omegazeng** [comment]
> Hello,
> 
> I have a similiar issue.
> My client is behind a corporate proxy and most of the kubectl commands work fine. But if I try to exec into a pod shell than I get following error:
> 
> ```
> kubectl exec -i -t -n default web-6798b588cf-4gzvq -c web "--" sh -c "clear; (bash || ash || sh)"
> error: unable to upgrade connection: error dialing backend: dial tcp: lookup .....eks.amazonaws.com on 10.58.194.16:53: no such host
> ```
> 
> So kubectl tries to resolute the public EKS domain with our internal DNS. Our internal DNS does not support resolution of public domains. It looks like the similiar root cause for me. Do you know if there is a workaround for that or will this bug be fixed soon?

use socks5
```
ssh -D 1337 -qCN [ssh server IP]
HTTPS_PROXY=socks://localhost:1337 kubectl get pods
```

**Comment by bagnaram** [comment]
> > Hello,
> > I have a similiar issue.
> > My client is behind a corporate proxy and most of the kubectl commands work fine. But if I try to exec into a pod shell than I get following error:
> > ```
> > kubectl exec -i -t -n default web-6798b588cf-4gzvq -c web "--" sh -c "clear; (bash || ash || sh)"
> > error: unable to upgrade connection: error dialing backend: dial tcp: lookup .....eks.amazonaws.com on 10.58.194.16:53: no such host
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So kubectl tries to resolute the public EKS domain with our internal DNS. Our internal DNS does not support resolution of public domains. It looks like the similiar root cause for me. Do you know if there is a workaround for that or will this bug be fixed soon?
> 
> use socks5
> 
> ```
> ssh -D 1337 -qCN [ssh server IP]
> HTTPS_PROXY=socks://localhost:1337 kubectl get pods
> ```

Unfortunately this will not work because kubectl utilizes UDP streaming which is not subject to SOCKS5 proxying

**Comment by terinjokes** [comment]
The implementation from #84205 has worked well for us, we've patched kubectl internally with it.

**Comment by leonK-BO** [comment]
> The implementation from #84205 has worked well for us, we've patched kubectl internally with it.

How did you patch kubectl internally with that? Could you provide a short how to?

**Comment by jtyr** [comment]
@leonK-BO You can try this:

```shell
git clone https://github.com/kubernetes/kubernetes.git
cd kubernetes
curl -s https://github.com/jtyr/kubernetes/commit/43deaf0ba43ee5d257418fd95e9a93c5d4afa0c5.diff | git apply -
go build -o ~/bin/kubectl ./cmd/kubectl
```

This is using a [patch](https://github.com/jtyr/kubernetes/commit/43deaf0ba43ee5d257418fd95e9a93c5d4afa0c5) that I have extracted from the #84205.

**Comment by leonK-BO** [comment]
> @leonK-BO You can try this:
> 
> ```shell
> git clone https://github.com/kubernetes/kubernetes.git
> cd kubernetes
> curl -s https://github.com/jtyr/kubernetes/commit/43deaf0ba43ee5d257418fd95e9a93c5d4afa0c5.diff | git apply -
> go build -o ~/bin/kubectl ./cmd/kubectl
> ```
> 
> This is using a [patch](https://github.com/jtyr/kubernetes/commit/43deaf0ba43ee5d257418fd95e9a93c5d4afa0c5) that I have extracted from the #84205.

Thanks for the short guide. I tried it and I think it worked (how can I verify that?). But I still get an error:

`kubectl exec -i -t -n lens-metrics prometheus-0 -c prometheus "--" sh -c "clear; (bash || ash || sh)"
Error from server (BadRequest): Upgrade request required`

Any idea what this can be?

**Comment by terinjokes** [comment]
That looks like a different error related to your environment, without this patch exec requests to a SOCKS5 proxy fail with `EOF`.

Are you setting the `HTTPS_PROXY` environment (or corresponding proxy setting in `~/.kube/config`) to include `socks5://`?

**Comment by leonK-BO** [comment]
> That looks like a different error related to your environment, without this patch exec requests to a SOCKS5 proxy fail with `EOF`.
> 
> Are you setting the `HTTPS_PROXY` environment (or corresponding proxy setting in `~/.kube/config`) to include `socks5://`?

Actually I do not set the HTTPS_PROXY in `~/.kube/config` I am doing that in my bash profile. But there I do not include `socks5://`. Should I include it there?

**Comment by terinjokes** [comment]
That's how Go (and `kubectl`) knows to use SOCKS5 mode instead of a [Web proxy server](https://en.wikipedia.org/wiki/Proxy_server#Web_proxy_servers).

**Comment by leonK-BO** [comment]
> That's how Go (and `kubectl`) knows to use SOCKS5 mode instead of a [Web proxy server](https://en.wikipedia.org/wiki/Proxy_server#Web_proxy_servers).

If I change the proxy from https to SOCKS5 in my bashprofile than all other CLI tools will be as well affected right? Is there a way to provide this SOCKS5 proxy only for kubectl?

**Comment by xens** [comment]
> > That's how Go (and `kubectl`) knows to use SOCKS5 mode instead of a [Web proxy server](https://en.wikipedia.org/wiki/Proxy_server#Web_proxy_servers).
> 
> If I change the proxy from https to SOCKS5 in my bashprofile than all other CLI tools will be as well affected right? Is there a way to provide this SOCKS5 proxy only for kubectl?

I believe that you can do something like this `alias kctl="env HTTPS_PROXY=socks5://proxy-url:port kubectl"`

**Comment by leonK-BO** [comment]
@xens thanks for the hint with `alias kctl="env HTTPS_PROXY=socks5://proxy-url:port kubectl"` it seems to work. But now I get following error message:
```
kctl exec -i -t -n kube-system aws-node-mxblz -c aws-node "--" sh -c "clear; (bash || ash || sh)"
   Unable to connect to the server: socks connect tcp 10.4.103.143:8080->YXZ.gr7.eu-central-1.eks.amazonaws.com:443: EOF
```

**Comment by terinjokes** [comment]
@leonK-BO Hmmm. I've gotten a similar error message (specifically "Unable to connect to the server: socks connect") when the SOCKS5 proxy server was in a bad state, and restarting it resolved the issue. Does your command work with kubectl for commands other than `exec` and `port-forward`? If so, what software are you using for the SOCKS5 proxy server?

**Comment by SamuAlfageme** [comment]
@leonK-BO in case it helps, I've just tested the `kubectl` patch from https://github.com/kubernetes/kubernetes/issues/58065#issuecomment-845071840 myself with the `proxy-url` set on my `KUBECONFIG` and it all worked out:

```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: [...]
    proxy-url: socks5://localhost:1337
    server: https://[...]:6443
    [...]
```

**Comment by leonK-BO** [comment]
@terinjokes & @SamuAlfageme Sorry for the late reply. I will try out in the next days again the commands and your proposals.
Just to make sure that I am here on the right path. The current corporate proxy was internally reachable via "http://xxxxxxxx" so this means that is a plain HTTP proxy right? In order to use a the applied kubectl patch for SOCKS5 I need the proxy to support SOCKS5 right? If that is true I need to first to find out if that protocol is supported...

And thanks again to all your support here!

**Comment by terinjokes** [comment]
You'll need to contact your IT department to know for sure.

**Comment by michailbrynard** [comment]
Another workaround, while waiting for https://github.com/kubernetes/kubernetes/pull/84205 to merge: use [Polipo](https://www.irif.fr/~jch/software/polipo/) to convert the SOCKS5 proxy into an HTTP proxy.
```
polipo socksParentProxy=proxy-url:port
env HTTPS_PROXY=http://127.0.0.1:8123 kubectl exec -it ubuntu -- bash
```

**Comment by patricknelson** [comment]
Quick update: The work to add SOCKS5 compatibility has moved from #84205 to #105632 and may be close to being completed soon.

**TL;DR / ELI5:** To get _everything_ to work smoothly right now, use an **HTTP proxy** over SOCKS because [`kubectl` and SOCKS don't seem to get along](https://github.com/kubernetes/kubernetes/issues/58065#issuecomment-603624335) when it comes to streaming stuff like `kubectl exec`. So, either use [`tinyproxy` on your proxy server](https://github.com/kubernetes/kubernetes/issues/58065#issuecomment-499580806) or opt for [Polipo combined with an existing SOCKS proxy](https://github.com/kubernetes/kubernetes/issues/58065#issuecomment-874860689). 

Also confirming that both of these approaches work fine with self-signed certificates when interacting with the API server as well (e.g. with `certificate-authority-data` in `.kube/config`). FWIW (people googling to get here): My use case was proxying a connection to GKE Autopilot, which uses a self-signed certificate and an endpoint pointing directly to an IP address. I wasn't sure how all of these proxies would affect SSL (particularly since my situation required finding a way to work around Netskope).

**Comment by xens** [comment]
FYI the work to fully support SOCKS proxies has been merged into master a while ago https://github.com/kubernetes/kubernetes/pull/105632.

As of today you can build `kubectl` from the master branch so patching is not needed anymore. It should be part of Kubernetes 1.24 that will be released in a month.

**Comment by xens** [comment]
The fix has been introduced in Kubernetes 1.24. /close
